File: homework/ba-nguyen/lession-3/kth-largest-element-in-an-array.ts
function swap(numbs: number[], i: number, j: number) {
  [numbs[i], numbs[j]] = [numbs[j], numbs[i]];
}

function partition(numbs: number[], left: number, right: number) {
  let boundary = left;
  const pivot = numbs[right];

  for (let i = left; i < right; i++) {
    if (numbs[i] < pivot) {
      swap(numbs, i, boundary);
      boundary++;
    }
  }
  swap(numbs, right, boundary);
  return boundary;
}

function quickSelect(numbs: number[], left: number, right: number, k: number) {
  const pivotIndex = partition(numbs, left, right);
  if (pivotIndex === k) return numbs[pivotIndex];
  if (pivotIndex < k) return quickSelect(numbs, pivotIndex + 1, right, k);
  return quickSelect(numbs, left, pivotIndex - 1, k);
}

function findKthLargest(numbs: number[], k: number) {
  return quickSelect(numbs, 0, numbs.length - 1, numbs.length - k);
}

File: homework/lesson3/kth-largest-element-in-an-array.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int len = nums.size();

        return quickSelect(nums, 0, len-1, len - k);
    }

private:
    int randomPivotIndex(int low, int high) {
        if (high == low)
            return low;

        return low + rand() % (high - low);
    }

    void swap(vector<int>& arr, int a, int b) {
        int term = arr[a];
        arr[a] = arr[b];
        arr[b] = term;

        return;
    }

    int partition(vector<int>& arr, int start, int end) {
        int pivotIndex = randomPivotIndex(start, end);
        int pivot = arr[pivotIndex];

        swap(arr, pivotIndex, end);

        int index = start;

        for(int x = start; x < end; ++x) {
            if (arr[x] < pivot) {
                swap(arr, index, x);
                index++;
            }
        }

        swap(arr, index, end);

        return index;
    }

    int quickSelect(vector<int>& arr, int start, int end, int k) {
        int pivotIndex = partition(arr, start, end);

        if (pivotIndex == k)
            return arr[pivotIndex];

        if (pivotIndex > k)
            return quickSelect(arr, start, pivotIndex - 1, k);

        return quickSelect(arr, pivotIndex + 1, end, k);
    }
};
